File: flutter/bicitaxi_conductor/lib/core/services/history_service.dart
import 'dart:async';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';

/// Model for a ride history entry
class RideHistoryEntry {
  final String rideId;
  final String role; // 'client' or 'driver'

  // Locations
  final double pickupLat;
  final double pickupLng;
  final String? pickupAddress;
  final double? dropoffLat;
  final double? dropoffLng;
  final String? dropoffAddress;

  // Parties
  final String clientUid;
  final String clientName;
  final String? driverUid;
  final String? driverName;

  // Timeline
  final DateTime createdAt;
  final DateTime? assignedAt;
  final DateTime? completedAt;
  final DateTime? cancelledAt;

  // Status
  final String status; // 'pending', 'assigned', 'completed', 'cancelled'

  RideHistoryEntry({
    required this.rideId,
    required this.role,
    required this.pickupLat,
    required this.pickupLng,
    this.pickupAddress,
    this.dropoffLat,
    this.dropoffLng,
    this.dropoffAddress,
    required this.clientUid,
    required this.clientName,
    this.driverUid,
    this.driverName,
    required this.createdAt,
    this.assignedAt,
    this.completedAt,
    this.cancelledAt,
    required this.status,
  });

  factory RideHistoryEntry.fromFirestore(Map<String, dynamic> data, String id) {
    return RideHistoryEntry(
      rideId: id,
      role: data['role'] ?? 'client',
      pickupLat: (data['pickupLat'] ?? 0).toDouble(),
      pickupLng: (data['pickupLng'] ?? 0).toDouble(),
      pickupAddress: data['pickupAddress'],
      dropoffLat: data['dropoffLat']?.toDouble(),
      dropoffLng: data['dropoffLng']?.toDouble(),
      dropoffAddress: data['dropoffAddress'],
      clientUid: data['clientUid'] ?? '',
      clientName: data['clientName'] ?? 'Cliente',
      driverUid: data['driverUid'],
      driverName: data['driverName'],
      createdAt: (data['createdAt'] as Timestamp?)?.toDate() ?? DateTime.now(),
      assignedAt: (data['assignedAt'] as Timestamp?)?.toDate(),
      completedAt: (data['completedAt'] as Timestamp?)?.toDate(),
      cancelledAt: (data['cancelledAt'] as Timestamp?)?.toDate(),
      status: data['status'] ?? 'pending',
    );
  }

  Map<String, dynamic> toFirestore() {
    return {
      'role': role,
      'pickupLat': pickupLat,
      'pickupLng': pickupLng,
      'pickupAddress': pickupAddress,
      'dropoffLat': dropoffLat,
      'dropoffLng': dropoffLng,
      'dropoffAddress': dropoffAddress,
      'clientUid': clientUid,
      'clientName': clientName,
      'driverUid': driverUid,
      'driverName': driverName,
      'createdAt': Timestamp.fromDate(createdAt),
      'assignedAt': assignedAt != null ? Timestamp.fromDate(assignedAt!) : null,
      'completedAt': completedAt != null
          ? Timestamp.fromDate(completedAt!)
          : null,
      'cancelledAt': cancelledAt != null
          ? Timestamp.fromDate(cancelledAt!)
          : null,
      'status': status,
    };
  }

  /// Display string for the ride date
  String get dateString {
    final now = DateTime.now();
    final diff = now.difference(createdAt);

    if (diff.inDays == 0) {
      return 'Hoy ${createdAt.hour.toString().padLeft(2, '0')}:${createdAt.minute.toString().padLeft(2, '0')}';
    } else if (diff.inDays == 1) {
      return 'Ayer';
    } else if (diff.inDays < 7) {
      return 'Hace ${diff.inDays} días';
    } else {
      return '${createdAt.day}/${createdAt.month}/${createdAt.year}';
    }
  }

  /// Status display text
  String get statusText {
    switch (status) {
      case 'completed':
        return 'Completado';
      case 'cancelled':
        return 'Cancelado';
      case 'assigned':
        return 'En curso';
      default:
        return 'Pendiente';
    }
  }
}

/// Service to manage ride history in Firestore
/// Each user has their own history collection at /users/{uid}/history/
class HistoryService {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;
  final FirebaseAuth _auth = FirebaseAuth.instance;

  String? get currentUserId => _auth.currentUser?.uid;

  /// Get reference to user's history collection
  CollectionReference<Map<String, dynamic>> _historyRef(String uid) {
    return _firestore.collection('users').doc(uid).collection('history');
  }

  /// Save or update history entry for a user
  Future<void> saveHistoryEntry(
    String uid,
    String rideId,
    RideHistoryEntry entry,
  ) async {
    await _historyRef(
      uid,
    ).doc(rideId).set(entry.toFirestore(), SetOptions(merge: true));
    print('📜 Saved history entry for user $uid, ride $rideId');
  }

  /// Update history entry status
  Future<void> updateHistoryStatus(
    String uid,
    String rideId,
    String status, {
    DateTime? assignedAt,
    DateTime? completedAt,
    DateTime? cancelledAt,
    String? driverUid,
    String? driverName,
  }) async {
    final updates = <String, dynamic>{'status': status};

    if (assignedAt != null)
      updates['assignedAt'] = Timestamp.fromDate(assignedAt);
    if (completedAt != null)
      updates['completedAt'] = Timestamp.fromDate(completedAt);
    if (cancelledAt != null)
      updates['cancelledAt'] = Timestamp.fromDate(cancelledAt);
    if (driverUid != null) updates['driverUid'] = driverUid;
    if (driverName != null) updates['driverName'] = driverName;

    await _historyRef(uid).doc(rideId).update(updates);
    print('📜 Updated history status for user $uid, ride $rideId: $status');
  }

  /// Watch user's ride history (stream)
  Stream<List<RideHistoryEntry>> watchHistory(String uid) {
    return _historyRef(uid)
        .orderBy('createdAt', descending: true)
        .limit(50)
        .snapshots()
        .map((snapshot) {
          return snapshot.docs.map((doc) {
            return RideHistoryEntry.fromFirestore(doc.data(), doc.id);
          }).toList();
        });
  }

  /// Get user's ride history (one-time fetch)
  Future<List<RideHistoryEntry>> getHistory(String uid) async {
    final snapshot = await _historyRef(
      uid,
    ).orderBy('createdAt', descending: true).limit(50).get();

    return snapshot.docs.map((doc) {
      return RideHistoryEntry.fromFirestore(doc.data(), doc.id);
    }).toList();
  }

  /// Create client history entry when request is created
  Future<void> createClientHistoryEntry({
    required String rideId,
    required String clientUid,
    required String clientName,
    required double pickupLat,
    required double pickupLng,
    String? pickupAddress,
    double? dropoffLat,
    double? dropoffLng,
    String? dropoffAddress,
  }) async {
    final entry = RideHistoryEntry(
      rideId: rideId,
      role: 'client',
      pickupLat: pickupLat,
      pickupLng: pickupLng,
      pickupAddress: pickupAddress,
      dropoffLat: dropoffLat,
      dropoffLng: dropoffLng,
      dropoffAddress: dropoffAddress,
      clientUid: clientUid,
      clientName: clientName,
      createdAt: DateTime.now(),
      status: 'pending',
    );

    await saveHistoryEntry(clientUid, rideId, entry);
  }

  /// Create driver history entry when ride is accepted
  Future<void> createDriverHistoryEntry({
    required String rideId,
    required String clientUid,
    required String clientName,
    required String driverUid,
    required String driverName,
    required double pickupLat,
    required double pickupLng,
    String? pickupAddress,
    double? dropoffLat,
    double? dropoffLng,
    String? dropoffAddress,
  }) async {
    final entry = RideHistoryEntry(
      rideId: rideId,
      role: 'driver',
      pickupLat: pickupLat,
      pickupLng: pickupLng,
      pickupAddress: pickupAddress,
      dropoffLat: dropoffLat,
      dropoffLng: dropoffLng,
      dropoffAddress: dropoffAddress,
      clientUid: clientUid,
      clientName: clientName,
      driverUid: driverUid,
      driverName: driverName,
      createdAt: DateTime.now(),
      assignedAt: DateTime.now(),
      status: 'assigned',
    );

    await saveHistoryEntry(driverUid, rideId, entry);
  }

  /// Mark ride as completed for both parties
  Future<void> markRideCompleted(
    String rideId,
    String clientUid,
    String? driverUid,
  ) async {
    final now = DateTime.now();

    // Update client's history
    await updateHistoryStatus(clientUid, rideId, 'completed', completedAt: now);

    // Update driver's history if driver exists
    if (driverUid != null) {
      await updateHistoryStatus(
        driverUid,
        rideId,
        'completed',
        completedAt: now,
      );
    }
  }

  /// Mark ride as cancelled
  Future<void> markRideCancelled(
    String rideId,
    String clientUid,
    String? driverUid,
  ) async {
    final now = DateTime.now();

    // Update client's history
    await updateHistoryStatus(clientUid, rideId, 'cancelled', cancelledAt: now);

    // Update driver's history if driver was assigned
    if (driverUid != null) {
      await updateHistoryStatus(
        driverUid,
        rideId,
        'cancelled',
        cancelledAt: now,
      );
    }
  }
}

File: ios/bicitaxi-conductor/bicitaxi-conductor/Services/HistoryService.swift
//
//  HistoryService.swift
//  bicitaxi
//
//  Service to manage ride history in Firestore
//  Each user has their own history collection at /users/{uid}/history/
//

import Foundation
import FirebaseFirestore
import FirebaseAuth
import Combine

/// Model for a ride history entry
struct RideHistoryEntry: Identifiable, Codable {
    let id: String  // Same as rideId
    let role: String  // "client" or "driver"
    
    // Locations
    let pickupLat: Double
    let pickupLng: Double
    let pickupAddress: String?
    let dropoffLat: Double?
    let dropoffLng: Double?
    let dropoffAddress: String?
    
    // Parties
    let clientUid: String
    let clientName: String
    let driverUid: String?
    let driverName: String?
    
    // Timeline
    let createdAt: Date
    let assignedAt: Date?
    let completedAt: Date?
    let cancelledAt: Date?
    
    // Status
    let status: String  // "pending", "assigned", "completed", "cancelled"
    
    /// Display string for the ride date
    var dateString: String {
        let now = Date()
        let diff = Calendar.current.dateComponents([.day], from: createdAt, to: now)
        
        if diff.day == 0 {
            let formatter = DateFormatter()
            formatter.dateFormat = "HH:mm"
            return "Hoy \(formatter.string(from: createdAt))"
        } else if diff.day == 1 {
            return "Ayer"
        } else if let days = diff.day, days < 7 {
            return "Hace \(days) días"
        } else {
            let formatter = DateFormatter()
            formatter.dateFormat = "dd/MM/yyyy"
            return formatter.string(from: createdAt)
        }
    }
    
    /// Status display text
    var statusText: String {
        switch status {
        case "completed": return "Completado"
        case "cancelled": return "Cancelado"
        case "assigned": return "En curso"
        default: return "Pendiente"
        }
    }
    
    /// Time display string
    var timeString: String {
        let formatter = DateFormatter()
        formatter.dateFormat = "HH:mm"
        return formatter.string(from: createdAt)
    }
}

/// Service to manage ride history in Firestore
class HistoryService: ObservableObject {
    private let db = Firestore.firestore()
    
    @Published var history: [RideHistoryEntry] = []
    @Published var isLoading = false
    
    private var listener: ListenerRegistration?
    
    var currentUserId: String? {
        Auth.auth().currentUser?.uid
    }
    
    /// Get reference to user's history collection
    private func historyRef(_ uid: String) -> CollectionReference {
        db.collection("users").document(uid).collection("history")
    }
    
    /// Fetch user's ride history
    func fetchHistory() async {
        guard let uid = currentUserId else {
            await MainActor.run { history = [] }
            return
        }
        
        await MainActor.run { isLoading = true }
        
        do {
            let snapshot = try await historyRef(uid)
                .order(by: "createdAt", descending: true)
                .limit(to: 50)
                .getDocuments()
            
            let entries = snapshot.documents.compactMap { doc -> RideHistoryEntry? in
                let data = doc.data()
                return RideHistoryEntry(
                    id: doc.documentID,
                    role: data["role"] as? String ?? "client",
                    pickupLat: data["pickupLat"] as? Double ?? 0,
                    pickupLng: data["pickupLng"] as? Double ?? 0,
                    pickupAddress: data["pickupAddress"] as? String,
                    dropoffLat: data["dropoffLat"] as? Double,
                    dropoffLng: data["dropoffLng"] as? Double,
                    dropoffAddress: data["dropoffAddress"] as? String,
                    clientUid: data["clientUid"] as? String ?? "",
                    clientName: data["clientName"] as? String ?? "Cliente",
                    driverUid: data["driverUid"] as? String,
                    driverName: data["driverName"] as? String,
                    createdAt: (data["createdAt"] as? Timestamp)?.dateValue() ?? Date(),
                    assignedAt: (data["assignedAt"] as? Timestamp)?.dateValue(),
                    completedAt: (data["completedAt"] as? Timestamp)?.dateValue(),
                    cancelledAt: (data["cancelledAt"] as? Timestamp)?.dateValue(),
                    status: data["status"] as? String ?? "pending"
                )
            }
            
            await MainActor.run {
                self.history = entries
                self.isLoading = false
            }
        } catch {
            print("⚠️ Error fetching history: \(error)")
            await MainActor.run {
                self.history = []
                self.isLoading = false
            }
        }
    }
    
    /// Watch user's history with real-time updates
    func watchHistory() {
        guard let uid = currentUserId else { return }
        
        listener?.remove()
        isLoading = true
        
        listener = historyRef(uid)
            .order(by: "createdAt", descending: true)
            .limit(to: 50)
            .addSnapshotListener { [weak self] snapshot, error in
                guard let self = self else { return }
                
                if let error = error {
                    print("⚠️ History listener error: \(error)")
                    self.isLoading = false
                    return
                }
                
                guard let documents = snapshot?.documents else {
                    self.history = []
                    self.isLoading = false
                    return
                }
                
                self.history = documents.compactMap { doc -> RideHistoryEntry? in
                    let data = doc.data()
                    return RideHistoryEntry(
                        id: doc.documentID,
                        role: data["role"] as? String ?? "client",
                        pickupLat: data["pickupLat"] as? Double ?? 0,
                        pickupLng: data["pickupLng"] as? Double ?? 0,
                        pickupAddress: data["pickupAddress"] as? String,
                        dropoffLat: data["dropoffLat"] as? Double,
                        dropoffLng: data["dropoffLng"] as? Double,
                        dropoffAddress: data["dropoffAddress"] as? String,
                        clientUid: data["clientUid"] as? String ?? "",
                        clientName: data["clientName"] as? String ?? "Cliente",
                        driverUid: data["driverUid"] as? String,
                        driverName: data["driverName"] as? String,
                        createdAt: (data["createdAt"] as? Timestamp)?.dateValue() ?? Date(),
                        assignedAt: (data["assignedAt"] as? Timestamp)?.dateValue(),
                        completedAt: (data["completedAt"] as? Timestamp)?.dateValue(),
                        cancelledAt: (data["cancelledAt"] as? Timestamp)?.dateValue(),
                        status: data["status"] as? String ?? "pending"
                    )
                }
                self.isLoading = false
            }
    }
    
    func stopWatching() {
        listener?.remove()
        listener = nil
    }
    
    // MARK: - Static write methods (used by RequestService)
    
    /// Save history entry for a user
    static func saveHistoryEntry(uid: String, rideId: String, data: [String: Any]) async {
        do {
            try await Firestore.firestore()
                .collection("users")
                .document(uid)
                .collection("history")
                .document(rideId)
                .setData(data, merge: true)
            print("📜 Saved history for user \(uid), ride \(rideId)")
        } catch {
            print("⚠️ Failed to save history: \(error)")
        }
    }
    
    /// Create client history entry when request is created
    static func createClientHistory(
        rideId: String,
        clientUid: String,
        clientName: String,
        pickupLat: Double,
        pickupLng: Double,
        pickupAddress: String?,
        dropoffLat: Double?,
        dropoffLng: Double?,
        dropoffAddress: String?
    ) async {
        let data: [String: Any] = [
            "role": "client",
            "pickupLat": pickupLat,
            "pickupLng": pickupLng,
            "pickupAddress": pickupAddress as Any,
            "dropoffLat": dropoffLat as Any,
            "dropoffLng": dropoffLng as Any,
            "dropoffAddress": dropoffAddress as Any,
            "clientUid": clientUid,
            "clientName": clientName,
            "createdAt": Timestamp(date: Date()),
            "status": "pending"
        ]
        await saveHistoryEntry(uid: clientUid, rideId: rideId, data: data)
    }
    
    /// Create driver history entry when ride is accepted
    static func createDriverHistory(
        rideId: String,
        clientUid: String,
        clientName: String,
        driverUid: String,
        driverName: String,
        pickupLat: Double,
        pickupLng: Double,
        pickupAddress: String?,
        dropoffLat: Double?,
        dropoffLng: Double?,
        dropoffAddress: String?
    ) async {
        let data: [String: Any] = [
            "role": "driver",
            "pickupLat": pickupLat,
            "pickupLng": pickupLng,
            "pickupAddress": pickupAddress as Any,
            "dropoffLat": dropoffLat as Any,
            "dropoffLng": dropoffLng as Any,
            "dropoffAddress": dropoffAddress as Any,
            "clientUid": clientUid,
            "clientName": clientName,
            "driverUid": driverUid,
            "driverName": driverName,
            "createdAt": Timestamp(date: Date()),
            "assignedAt": Timestamp(date: Date()),
            "status": "assigned"
        ]
        await saveHistoryEntry(uid: driverUid, rideId: rideId, data: data)
        
        // Also update client's history with driver info
        let clientUpdate: [String: Any] = [
            "driverUid": driverUid,
            "driverName": driverName,
            "assignedAt": Timestamp(date: Date()),
            "status": "assigned"
        ]
        await saveHistoryEntry(uid: clientUid, rideId: rideId, data: clientUpdate)
    }
    
    /// Mark ride completed for both parties
    static func markRideCompleted(rideId: String, clientUid: String, driverUid: String?) async {
        print("🔍 markRideCompleted called: rideId=\(rideId), clientUid=\(clientUid), driverUid=\(driverUid ?? "nil")")
        
        let update: [String: Any] = [
            "completedAt": Timestamp(date: Date()),
            "status": "completed"
        ]
        
        // Only update client history if we have a valid uid
        if !clientUid.isEmpty {
            await saveHistoryEntry(uid: clientUid, rideId: rideId, data: update)
        } else {
            print("⚠️ Cannot update client history: clientUid is empty")
        }
        
        if let driverUid = driverUid, !driverUid.isEmpty {
            await saveHistoryEntry(uid: driverUid, rideId: rideId, data: update)
        }
    }
    
    /// Mark ride cancelled
    static func markRideCancelled(rideId: String, clientUid: String, driverUid: String?) async {
        let update: [String: Any] = [
            "cancelledAt": Timestamp(date: Date()),
            "status": "cancelled"
        ]
        await saveHistoryEntry(uid: clientUid, rideId: rideId, data: update)
        if let driverUid = driverUid {
            await saveHistoryEntry(uid: driverUid, rideId: rideId, data: update)
        }
    }
    
    deinit {
        listener?.remove()
    }
}
